File: plugins/plugin-discord/rust/src/actions/get_user_info.rs
//! Get user info action

use async_trait::async_trait;

use super::{ActionContext, ActionResult, DiscordAction};
use crate::error::Result;
use crate::DiscordService;

/// Action to get detailed information about a Discord user
pub struct GetUserInfoAction;

#[async_trait]
impl DiscordAction for GetUserInfoAction {
    fn name(&self) -> &str {
        "GET_USER_INFO"
    }

    fn description(&self) -> &str {
        "Get detailed information about a Discord user including their roles, \
        join date, and permissions."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "USER_INFO",
            "WHO_IS",
            "ABOUT_USER",
            "USER_DETAILS",
            "MEMBER_INFO",
            "CHECK_USER",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context
            .message
            .get("source")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        Ok(source == "discord")
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &DiscordService,
    ) -> Result<ActionResult> {
        let text = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        // Parse user identifier from message
        let user_identifier = service.extract_user_identifier(text).await?;

        let user_identifier = match user_identifier {
            Some(id) => id,
            None => {
                return Ok(ActionResult::failure(
                    "I couldn't understand which user you want information about. \
                    Please specify a username or mention.",
                ))
            }
        };

        let guild_id = match &context.guild_id {
            Some(id) => id,
            None => {
                return Ok(ActionResult::failure(
                    "I can only get user info in a server, not in DMs.",
                ))
            }
        };

        // Get user info
        let user_info = service.get_member_info(guild_id, &user_identifier).await?;

        let user_info = match user_info {
            Some(info) => info,
            None => {
                return Ok(ActionResult::failure(format!(
                    "I couldn't find a user with the identifier \"{}\" in this server.",
                    user_identifier
                )))
            }
        };

        // Format the response
        let formatted = service.format_user_info(&user_info);

        let user_id = user_info.get("id").and_then(|i| i.as_str()).unwrap_or("");
        let username = user_info
            .get("username")
            .and_then(|u| u.as_str())
            .unwrap_or("");

        Ok(ActionResult::success_with_data(
            formatted,
            serde_json::json!({
                "user_id": user_id,
                "username": username,
            }),
        ))
    }
}

File: plugins/plugin-discord/python/elizaos_plugin_discord/actions/get_user_info.py
"""Get user info action."""

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from elizaos_plugin_discord.actions import ActionContext, ActionResult
    from elizaos_plugin_discord.service import DiscordService


class GetUserInfoAction:
    """Action to get detailed information about a Discord user."""

    @property
    def name(self) -> str:
        return "GET_USER_INFO"

    @property
    def description(self) -> str:
        return (
            "Get detailed information about a Discord user including their roles, "
            "join date, and permissions."
        )

    @property
    def similes(self) -> list[str]:
        return [
            "USER_INFO",
            "WHO_IS",
            "ABOUT_USER",
            "USER_DETAILS",
            "MEMBER_INFO",
            "CHECK_USER",
        ]

    async def validate(self, context: "ActionContext") -> bool:
        """Validate the action can be executed."""
        source = context.message.get("source")
        return isinstance(source, str) and source == "discord" and context.guild_id is not None

    async def handler(
        self,
        context: "ActionContext",
        service: "DiscordService",
    ) -> "ActionResult":
        """Execute the action."""
        from elizaos_plugin_discord.actions import ActionResult

        if context.guild_id is None:
            return ActionResult.failure_result(
                "I can only look up user info inside a server (not in DMs)."
            )

        content = context.message.get("content", {})
        text = content.get("text", "") if isinstance(content, dict) else ""

        # Parse user identifier from message
        user_identifier = await service.extract_user_identifier(text)
        if not user_identifier:
            return ActionResult.failure_result(
                "I couldn't understand which user you want information about. "
                "Please specify a username or mention."
            )

        # Get user info
        user_info = await service.get_member_info(context.guild_id, user_identifier)
        if not user_info:
            return ActionResult.failure_result(
                f'I couldn\'t find a user with the identifier "{user_identifier}" in this server.'
            )

        # Format the response
        formatted = service.format_user_info(user_info)

        return ActionResult.success_result(
            formatted,
            {
                "user_id": user_info.get("id", ""),
                "username": user_info.get("username", ""),
            },
        )
